src/OpenTelemetry/Trace/ParentBasedSampler.cs

```diff
@@ -37,6 +47,41 @@ public ParentBasedSampler(Sampler rootSampler)
             this.Description = $"ParentBased{{{rootSampler.Description}}}";
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ParentBasedSampler"/> class with ability to delegate
+        /// sampling decision to one of the inner samplers provided.
+        /// </summary>
+        /// <param name="rootSampler">The <see cref="Sampler"/> to be called for root span/activity.</param>
+        /// <param name="remoteParentSampled">
+        /// A <see cref="Sampler"/> to delegate sampling decision to in case of
+        /// remote parent (<see cref="ActivityContext.IsRemote"/> == true) with <see cref="ActivityTraceFlags.Recorded"/> flag == true.
+        /// </param>
+        /// <param name="remoteParentNotSampled">
+        /// A <see cref="Sampler"/> to delegate sampling decision to in case of
+        /// remote parent (<see cref="ActivityContext.IsRemote"/> == true) with <see cref="ActivityTraceFlags.Recorded"/> flag == false.
+        /// </param>
+        /// <param name="localParentSampled">
+        /// A <see cref="Sampler"/> to delegate sampling decision to in case of
+        /// local parent (<see cref="ActivityContext.IsRemote"/> == false) with <see cref="ActivityTraceFlags.Recorded"/> flag == true.
+        /// </param>
+        /// <param name="localParentNotSampled">
+        /// A <see cref="Sampler"/> to delegate sampling decision to in case of
+        /// local parent (<see cref="ActivityContext.IsRemote"/> == false) with <see cref="ActivityTraceFlags.Recorded"/> flag == false.
+        /// </param>
+        public ParentBasedSampler(
+            Sampler rootSampler,
+            Sampler remoteParentSampled,
+            Sampler remoteParentNotSampled,
+            Sampler localParentSampled,
+            Sampler localParentNotSampled)
+            : this(rootSampler)
```

Comment:
The other language implementations I've seen for this sampler always have the fields populated, eg. `remoteParentSampled ` would be set to an instance of `AlwaysOnSampler`. This would simplify `ShouldSample` as null checks are no longer required.

* [Ruby](https://github.com/open-telemetry/opentelemetry-ruby/blob/b3fd39fa202a0168f7c48304a8ba105d3f39bc1d/sdk/lib/opentelemetry/sdk/trace/samplers.rb#L67-L75)
* [Go](https://github.com/open-telemetry/opentelemetry-go/blob/c066f15ed74fec52c4755014bfcf2bdfb8e0e661/sdk/trace/sampling.go#L170-L175)

Reply:
Based on those two implementations and my understanding of the specification ("Optional parameters"), I would say users should be allowed to specify a single override sampler, and not require that all 4 parameters or passed.

Reply:
Thx! I also thought about this, but wanted to avoid creating additional instances of AlwaysOn/AlwaysOff samplers and the delegating, assuming they will be overridden quite rarely. But not really sure how big of an issue it would be.

I do agree it makes it inconvenient to require passing all 4 in case someone only needed on (or likely two). Perhaps a better api like below?
````csharp
new ParentBasedSampler(new AlwaysOnSampler()) { RemoteParentNotSampled = new AlwaysOnSampler() }
```

Reply:
> Thx! I also thought about this, but wanted to avoid creating additional instances of AlwaysOn/AlwaysOff samplers and the delegating, assuming they will be overridden quite rarely. But not really sure how big of an issue it would be.

Fair point.

> I do agree it makes it inconvenient to require passing all 4 in case someone only needed one (or likely two). Perhaps a better api like below?

Using properties would allow for the samplers to be changed after instantiation. Maybe optional parameters instead?

Reply:
Tricky API. What about a builder? Something like...
```csharp
new ParentBasedSamplerBuilder()
.AddSampler(target: ParentBasedSamplerTarget.RemoteSpan, whenParentSampled: true, sampler: new AlwaysOnSampler())
.AddSampler(target: ParentBasedSamplerTarget.LocalSpan, whenParentSampled: false, sampler: new AlwaysOffSampler())
.Build();
```
🤷

Reply:
Yeah we need init-only properties 😄 

I was also thinking of doing "options":
```csharp
new ParentBasedSampler(
    new AlwaysOnSampler(), // root
    new ParentBasedSamplerOptions
    {
        RemoteParentNotSampled = new AlwaysOnSampler()
    }
);
```

Build I like the builder too.

We should probably also add ability to customize link-sampling behavior through these while we're at it.

Reply:
@mbakalov Are you still exploring using the Options way or Builder way?

Reply:
@cijothomas - I now think maybe the current version I have is good enough?
```csharp
public ParentBasedSampler(Sampler rootSampler)
{
}

public ParentBasedSampler(
	Sampler rootSampler,
	Sampler remoteParentSampled = null,
	Sampler remoteParentNotSampled = null,
	Sampler localParentSampled = null,
	Sampler localParentNotSampled = null)
	: this(rootSampler)
{
}
```

Allows to do this:
```csharp
var pb = new ParentBasedSampler(new AlwaysOnSampler(), localParentNotSampled: new AlwaysOnSampler());
```

Which seems ok?

Java has the same nullable parameters constructor and also has the builder, but I think that's probably because you can't do named arguments for optional parameters in Java.

Reply:
This looks good enough.

Reply:
@johnduhart @CodeBlanch Please comment if you see concerns with this. This looks reasonable to me.

Reply:
Should we add the default samplers into the XML comments? Eg:
```csharp
        /// <param name="remoteParentNotSampled">
        /// A <see cref="Sampler"/> to delegate sampling decision to in case of
        /// remote parent (<see cref="ActivityContext.IsRemote"/> == true) with <see cref="ActivityTraceFlags.Recorded"/> flag == false.
        /// Default: <see cref="AlwaysOffSampler" />.
        /// </param>
```

Reply:
Wouldn't hurt - added!